File: dotnet/Visualizer/ViewControllers/VisualizerViewController.cs
using System;
using System.Threading.Tasks;
using AppKit;
using CoreFoundation;
using CoreGraphics;
using Foundation;
using Lego.Client;
using Lego.Core;
using Lego.Core.Description;
using Maths;
using Metal;
using MetalKit;
using Visualizer.Rendering.Car;

namespace Visualizer.ViewControllers
{
    public class VisualizerViewController : NSViewController, IRotationProvider, ILegoCarStateProvider
    {
        public event EventHandler OnDisconnect;
        
        private readonly LegoCarClient _client = new LegoCarClient(NSHost.Current.Name);
        private readonly IMTKViewDelegate _renderer;
        private readonly MTKView _mtkView;
        private readonly NSButton _disconnectButton;
        private readonly NSTextField _xMotionField;
        private readonly NSTextField _yMotionField;

        public VisualizerViewController()
        {
            var device = MTLDevice.SystemDefault;
            if (device == null)
            {
                throw new Exception("Metal is not supported on this device");
            }

            _mtkView = new MTKView(new CGRect(), device)
            {
                ColorPixelFormat = MTLPixelFormat.BGRA8Unorm,
                DepthStencilPixelFormat = MTLPixelFormat.Depth32Float
            };
            _renderer = new CarRenderer(_mtkView, this);
            _mtkView.Delegate = _renderer;
            
            _disconnectButton = new NSButton
            {
                Title = "Disconnect",
                Cell =
                {
                    Bezeled = true
                }
            };
            _disconnectButton.Activated += DisconnectButtonClicked;

            var xMotionLabel = new NSTextField {StringValue = "X-motion", Editable = false};
            _xMotionField = new NSTextField { Editable = false, StringValue = "     " };
            var yMotionLabel = new NSTextField {StringValue = "Y-motion", Editable = false};
            _yMotionField = new NSTextField { Editable = false, StringValue = "     " };
            

            var numberView = new NSStackView {Orientation = NSUserInterfaceLayoutOrientation.Horizontal}
                .WithArrangedSubviews(xMotionLabel, _xMotionField, yMotionLabel, _yMotionField);
            
            View = new NSView()
                .WithSubview(_disconnectButton, (c, p) => new[]
                {
                    c.TopAnchor.ConstraintEqualToAnchor(p.TopAnchor, 20),
                    c.CenterXAnchor.ConstraintEqualToAnchor(p.CenterXAnchor)
                })
                .WithSubview(_mtkView, (c, p) => new []
                {
                    c.TopAnchor.ConstraintEqualToAnchor(_disconnectButton.BottomAnchor),
                    c.LeadingAnchor.ConstraintEqualToAnchor(p.LeadingAnchor),
                    c.TrailingAnchor.ConstraintEqualToAnchor(p.TrailingAnchor)
                })
                .WithSubview(numberView, (c, p) => new[]
                {
                    c.TopAnchor.ConstraintEqualToAnchor(_mtkView.BottomAnchor),
                    c.LeadingAnchor.ConstraintEqualToAnchor(p.LeadingAnchor),
                    c.TrailingAnchor.ConstraintEqualToAnchor(p.TrailingAnchor),
                    c.BottomAnchor.ConstraintEqualToAnchor(p.BottomAnchor)
                })
                ;
            _client.DidUpdate = DidUpdate;
            _client.Disconnected = _ => DisconnectAsync();
        }

        private Task DidUpdate(LegoCarClient client, LegoCarState state)
        {
            DispatchQueue.MainQueue.DispatchAsync(() =>
            {
                _xMotionField.IntValue = state.Motion.X;
                _yMotionField.IntValue = state.Motion.Y;
            });
            return Task.CompletedTask;
        }

        public async Task ConnectAsync(string host, int port)
        {
            await _client.ConnectAsync(host, port);
        }

        protected override void Dispose(bool disposing)
        {
            if (disposing)
            {
                _disconnectButton.Activated -= DisconnectButtonClicked;
                _client.DisconnectAsync().Wait(1000);
                _client.Dispose();
                _renderer.Dispose();
                _mtkView.Dispose();
            }
            base.Dispose(disposing);
        }

        private async Task DisconnectAsync()
        {
            await _client.DisconnectAsync();
            OnDisconnect?.Invoke(this, EventArgs.Empty);
        }

        private async void DisconnectButtonClicked(object sender, EventArgs e)
        {
            await DisconnectAsync();
        }

        public LegoCarState GetState()
        {
            return _client?.GetState() ?? new LegoCarState();
        }

        public LegoCarDescriptor GetCarDescriptor() => LegoCarDescriptor.RaceCar();

        public Double3 GetEulerAngles() => _client?.GetEulerAngles() ?? Double3.Zero;

        public Quatd GetQuaternion() => _client?.GetQuaternion() ?? new Quatd(0, 0, 0);

    }
}

File: apple/Visualizer/Visualizer/Views/VisualizerViewController.swift
import Foundation
import AppKit
import MetalKit
import Metal

class VisualizerViewController: NSViewController {

    let client : LegoCarClient
    let mtkView: MTKView
    let renderer: CarRenderer
    var delegate: VisualizerViewControllerDelegate?

    required init(coder: NSCoder) {
        fatalError("No nibs here")
    }

    init() {
        guard let device = MTLCreateSystemDefaultDevice() else {
            fatalError("Metal not supported")
        }

        mtkView = MTKView(frame: CGRect(), device: device)
        mtkView.colorPixelFormat = .bgra8Unorm
        mtkView.depthStencilPixelFormat = .depth32Float

        client = LegoCarClient()
        renderer = CarRenderer(view: mtkView, stateProvider: client)
        super.init(nibName: nil, bundle: nil)

        let disconnectButton = NSButton(title: "Disconnect", target: self, action: #selector(disconnectClicked))

        view = NSView()
        .withSubview(mtkView, constraints: {c, p in [
            c.centerXAnchor.constraint(equalTo: p.centerXAnchor),
            c.centerYAnchor.constraint(equalTo: p.centerYAnchor),
            c.widthAnchor.constraint(equalTo: p.widthAnchor, multiplier: 0.8),
            c.heightAnchor.constraint(equalTo: p.heightAnchor, multiplier: 0.8)
        ]})
        .withSubview(disconnectButton, constraints: { c, p in [
            c.topAnchor.constraint(equalTo: p.topAnchor, constant: 20),
            c.centerXAnchor.constraint(equalTo: p.centerXAnchor)
        ]})
    }

    @objc
    func disconnectClicked() {
        print("Disconnect clicked")
        delegate?.visualizerViewController(self, disconnected: "by user")
    }

    func connect(host: String, port: Int) {

    }
}

protocol VisualizerViewControllerDelegate {
    func visualizerViewController(_ controller: VisualizerViewController, disconnected: String)
}
